File: app/src/main/java/com/zuk0/gaijinsmash/riderz/data/local/converter/Converters.java
package com.zuk0.gaijinsmash.riderz.data.local.converter;

import android.arch.persistence.room.TypeConverter;
import android.arch.persistence.room.TypeConverters;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.zuk0.gaijinsmash.riderz.data.local.entity.Favorite;
import com.zuk0.gaijinsmash.riderz.data.local.entity.bsa_response.Bsa;
import com.zuk0.gaijinsmash.riderz.data.local.entity.etd_response.Estimate;
import com.zuk0.gaijinsmash.riderz.data.local.entity.etd_response.EtdXmlResponse;
import com.zuk0.gaijinsmash.riderz.data.local.entity.station_response.Station;
import com.zuk0.gaijinsmash.riderz.data.local.entity.trip_response.Fare;
import com.zuk0.gaijinsmash.riderz.data.local.entity.trip_response.Fares;
import com.zuk0.gaijinsmash.riderz.data.local.entity.trip_response.Leg;
import com.zuk0.gaijinsmash.riderz.data.local.entity.trip_response.Trip;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;

import static com.zuk0.gaijinsmash.riderz.data.local.entity.Favorite.Priority.OFF;
import static com.zuk0.gaijinsmash.riderz.data.local.entity.Favorite.Priority.ON;

public class Converters {

    @TypeConverter
    public static ArrayList<String> toStringList(String value) {
        if(value ==null) {
            return new ArrayList<String>();
        } else {
            Type listType = new TypeToken<ArrayList<String>>() {}.getType();
            return new Gson().fromJson(value, listType);
        }
    }

    @TypeConverter
    public static String arrayListToString(ArrayList<String> list) {
        if(list == null) {
            return "";
        } else {
            Gson gson = new Gson();
            return gson.toJson(list);
        }
    }

    @TypeConverter
    public static int toInt(Favorite.Priority priority) {
        if(priority == OFF) {
            return 0;
        }
        if(priority == ON) {
            return 1;
        }
        return 0;
    }

    @TypeConverter
    public static Favorite.Priority toPriority(int value) {
        if(value == 0) {
            return OFF;
        } else if (value == 1) {
            return ON;
        } else {
            throw new IllegalArgumentException("Wrong argument. Value must be either 0 or 1");
        }
    }

    //todo: convert to ArraySet instead of HashSet for increased performance
    @TypeConverter
    public static HashSet<String> toHashType(String value) {
        if(value == null) {
            return new HashSet<>();
        } else {
            Type listType = new TypeToken<HashSet<String>>() {}.getType();
            return new Gson().fromJson(value, listType);
        }
    }

    //todo: convert to ArraySet instead of HashSet for increased performance
    @TypeConverter
    public static String hashSetToString(HashSet<String> set) {
        if(set == null) {
            return "";
        } else {
            Gson gson = new Gson();
            return gson.toJson(set);
        }
    }

    @TypeConverter
    public static Bsa toBSAType(String value) {
        if(value == null) {
            return new Bsa();
        } else {
            Type listType = new TypeToken<Bsa>() {}.getType();
            return new Gson().fromJson(value, listType);
        }
    }

    @TypeConverter
    public static String bsaToString(Bsa bsa) {
        if(bsa == null) {
            return "";
        } else {
            Gson gson = new Gson();
            return gson.toJson(bsa);
        }
    }

    @TypeConverter
    public static List<Bsa> toBSAListType(String value) {
        if(value == null) {
            return new ArrayList<Bsa>();
        } else {
            Type listType = new TypeToken<List<Bsa>>() {}.getType();
            return new Gson().fromJson(value, listType);
        }
    }

    @TypeConverter
    public static String bsaListToString(List<Bsa> bsaList) {
        if(bsaList == null) {
            return "";
        } else {
            Gson gson = new Gson();
            return gson.toJson(bsaList);
        }
    }

    @TypeConverter
    public static String stationToString(Station station) {
        if(station == null) {
            return "";
        } else {
            Gson gson = new Gson();
            return gson.toJson(station);
        }
    }

    @TypeConverter
    public static Station stringToStation(String value) {
        if(value == null || value.equals("")) {
            return new Station();
        } else {
            Type listType = new TypeToken<Station>() {}.getType();
            return new Gson().fromJson(value, listType);
        }
    }

    @TypeConverter
    public static Trip stringToTrip(String value) {
        if(value == null || value.equals("")) {
            return new Trip();
        } else {
            Type listType = new TypeToken<Trip>() {}.getType();
            return new Gson().fromJson(value, listType);
        }
    }

    @TypeConverter
    public static String TripToString(Trip trip) {
        if(trip == null) {
            return "";
        } else {
            Gson gson = new Gson();
            return gson.toJson(trip);
        }
    }

    @TypeConverter
    public static Fare stringToFare(String value) {
        if(value == null || value.equals("")) {
            return new Fare();
        } else {
            Type listType = new TypeToken<Fare>() {}.getType();
            return new Gson().fromJson(value, listType);
        }
    }

    @TypeConverter
    public static String fareToString(Fare fare) {
        if(fare == null) {
            return "";
        } else {
            Gson gson = new Gson();
            return gson.toJson(fare);
        }
    }

    @TypeConverter
    public static Fares stringToFares(String value) {
        if(value == null || value.equals("")) {
            return new Fares();
        } else {
            Type listType = new TypeToken<Fares>() {}.getType();
            return new Gson().fromJson(value, listType);
        }
    }

    @TypeConverter
    public static String faresToString(Fares fares) {
        if(fares == null) {
            return "";
        } else {
            Gson gson = new Gson();
            return gson.toJson(fares);
        }
    }

    @TypeConverter
    public static Leg stringToLeg(String value) {
        if(value == null || value.equals("")) {
            return new Leg();
        } else {
            Type listType = new TypeToken<Leg>() {}.getType();
            return new Gson().fromJson(value, listType);
        }
    }

    @TypeConverter
    public static String legToString(Leg leg) {
        if(leg == null) {
            return "";
        } else {
            Gson gson = new Gson();
            return gson.toJson(leg);
        }
    }

    @TypeConverter
    public static List<Leg> stringToLegList(String value) {
        if(value == null || value.equals("")) {
            return new ArrayList<>();
        } else {
            Type listType = new TypeToken<List<Leg>>() {}.getType();
            return new Gson().fromJson(value, listType);
        }
    }

    @TypeConverter
    public static String legListToString(List<Leg> legList) {
        if(legList == null) {
            return "";
        } else {
            Gson gson = new Gson();
            return gson.toJson(legList);
        }
    }

    @TypeConverter
    public static List<Estimate> stringToEstimateList(String value) {
        if(value == null || value.equals("")){
            return new ArrayList<>();
        } else {
            Type listType = new TypeToken<List<Estimate>>() {}.getType();
            return new Gson().fromJson(value, listType);
        }
    }

    @TypeConverter
    public static String estimateListToString(List<Estimate> estimateList) {
        if(estimateList == null) {
            return "";
        } else {
            Gson gson = new Gson();
            return gson.toJson(estimateList);
        }
    }

    @TypeConverter
    public static List<Fare> stringToFareList(String value) {
        if(value == null || value.equals("")){
            return new ArrayList<>();
        } else {
            Type listType = new TypeToken<List<Fare>>() {}.getType();
            return new Gson().fromJson(value, listType);
        }
    }

    @TypeConverter
    public static String fareListToString(List<Fare> fareList) {
        if(fareList == null) {
            return "";
        } else {
            Gson gson = new Gson();
            return gson.toJson(fareList);
        }
    }

}

File: app/src/main/java/com/zuk0/gaijinsmash/riderz/data/local/room/converter/Converters.kt
package com.zuk0.gaijinsmash.riderz.data.local.room.converter

import android.arch.persistence.room.TypeConverter

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.zuk0.gaijinsmash.riderz.data.local.entity.Favorite
import com.zuk0.gaijinsmash.riderz.data.local.entity.bsa_response.Bsa
import com.zuk0.gaijinsmash.riderz.data.local.entity.etd_response.Estimate
import com.zuk0.gaijinsmash.riderz.data.local.entity.station_response.Station
import com.zuk0.gaijinsmash.riderz.data.local.entity.trip_response.Fare
import com.zuk0.gaijinsmash.riderz.data.local.entity.trip_response.Fares
import com.zuk0.gaijinsmash.riderz.data.local.entity.trip_response.Leg
import com.zuk0.gaijinsmash.riderz.data.local.entity.trip_response.Trip

import java.util.ArrayList
import java.util.HashSet

import com.zuk0.gaijinsmash.riderz.data.local.entity.Favorite.Priority.OFF
import com.zuk0.gaijinsmash.riderz.data.local.entity.Favorite.Priority.ON
import java.sql.Timestamp

class Converters {

    @TypeConverter
    fun toStringList(value: String?): ArrayList<String> {
        if (value.equals("") || value == null) {
            return ArrayList()
        } else {
            val listType = object : TypeToken<ArrayList<String>>() {}.type
            return Gson().fromJson(value, listType)
        }
    }

    @TypeConverter
    fun arrayListToString(list: ArrayList<String>?): String {
        if (list == null) {
            return ""
        } else {
            val gson = Gson()
            return gson.toJson(list)
        }
    }

    @TypeConverter
    fun favoriteToString(favorite: Favorite) : String {
        if(false) {
            return ""
        } else {
            val gson = Gson()
            return gson.toJson(favorite)
        }
    }

    @TypeConverter
    fun stringToFavorite(value: String) : Favorite {
        if(value.equals("") ) {
            return Favorite()
        } else {
            val listType = object : TypeToken<Favorite>() {}.type
            return Gson().fromJson(value, listType)
        }
    }

    @TypeConverter
    fun toInt(priority: Favorite.Priority): Int {
        if (priority == OFF) {
            return 0
        }
        return if (priority == ON) {
            1
        } else 0
    }

    @TypeConverter
    fun toPriority(value: Int): Favorite.Priority {
        return if (value.equals(0)) {
            OFF
        } else if (value.equals(1)) {
            ON
        } else {
            throw IllegalArgumentException("Wrong argument. Value must be either 0 or 1")
        }
    }

    //todo: convert to ArraySet instead of HashSet for increased performance
    @TypeConverter
    fun toHashType(value: String?): HashSet<String> {
        if (value == "") {
            return HashSet()
        } else {
            val listType = object : TypeToken<HashSet<String>>() {

            }.type
            return Gson().fromJson(value, listType)
        }
    }

    //todo: convert to ArraySet instead of HashSet for increased performance
    @TypeConverter
    fun hashSetToString(set: HashSet<String>?): String {
        if (set == null) {
            return ""
        } else {
            val gson = Gson()
            return gson.toJson(set)
        }
    }

    @TypeConverter
    fun toBSAType(value: String?): Bsa {
        if (value == "") {
            return Bsa()
        } else {
            val listType = object : TypeToken<Bsa>() {

            }.type
            return Gson().fromJson(value, listType)
        }
    }

    @TypeConverter
    fun bsaToString(bsa: Bsa?): String {
        if (bsa == null) {
            return ""
        } else {
            val gson = Gson()
            return gson.toJson(bsa)
        }
    }

    @TypeConverter
    fun toBSAListType(value: String?): List<Bsa> {
        if (value == "") {
            return ArrayList()
        } else {
            val listType = object : TypeToken<List<Bsa>>() {

            }.type
            return Gson().fromJson(value, listType)
        }
    }

    @TypeConverter
    fun bsaListToString(bsaList: List<Bsa>?): String {
        if (bsaList == null) {
            return ""
        } else {
            val gson = Gson()
            return gson.toJson(bsaList)
        }
    }

    @TypeConverter
    fun stationToString(station: Station?): String {
        if (station == null) {
            return ""
        } else {
            val gson = Gson()
            return gson.toJson(station)
        }
    }

    @TypeConverter
    fun stringToStation(value: String?): Station {
        if (value == null || value == "") {
            return Station()
        } else {
            val listType = object : TypeToken<Station>() {

            }.type
            return Gson().fromJson(value, listType)
        }
    }

    @TypeConverter
    fun stringToTrip(value: String?): Trip {
        if (value == null || value == "") {
            return Trip()
        } else {
            val listType = object : TypeToken<Trip>() {

            }.type
            return Gson().fromJson(value, listType)
        }
    }

    @TypeConverter
    fun TripToString(trip: Trip?): String {
        if (trip == null) {
            return ""
        } else {
            val gson = Gson()
            return gson.toJson(trip)
        }
    }

    @TypeConverter
    fun stringToFare(value: String?): Fare {
        if (value == null || value == "") {
            return Fare()
        } else {
            val listType = object : TypeToken<Fare>() {

            }.type
            return Gson().fromJson(value, listType)
        }
    }

    @TypeConverter
    fun fareToString(fare: Fare?): String {
        if (fare == null) {
            return ""
        } else {
            val gson = Gson()
            return gson.toJson(fare)
        }
    }

    @TypeConverter
    fun stringToFares(value: String?): Fares {
        if (value == null || value == "") {
            return Fares()
        } else {
            val listType = object : TypeToken<Fares>() {

            }.type
            return Gson().fromJson(value, listType)
        }
    }

    @TypeConverter
    fun faresToString(fares: Fares?): String {
        if (fares == null) {
            return ""
        } else {
            val gson = Gson()
            return gson.toJson(fares)
        }
    }

    @TypeConverter
    fun stringToLeg(value: String?): Leg {
        if (value == null || value == "") {
            return Leg()
        } else {
            val listType = object : TypeToken<Leg>() {

            }.type
            return Gson().fromJson(value, listType)
        }
    }

    @TypeConverter
    fun legToString(leg: Leg?): String {
        if (leg == null) {
            return ""
        } else {
            val gson = Gson()
            return gson.toJson(leg)
        }
    }

    @TypeConverter
    fun stringToLegList(value: String?): List<Leg> {
        if (value == null || value == "") {
            return ArrayList()
        } else {
            val listType = object : TypeToken<List<Leg>>() {

            }.type
            return Gson().fromJson(value, listType)
        }
    }

    @TypeConverter
    fun legListToString(legList: List<Leg>?): String {
        if (legList == null) {
            return ""
        } else {
            val gson = Gson()
            return gson.toJson(legList)
        }
    }

    @TypeConverter
    fun stringToEstimateList(value: String?): List<Estimate> {
        if (value == null || value == "") {
            return ArrayList()
        } else {
            val listType = object : TypeToken<List<Estimate>>() {

            }.type
            return Gson().fromJson(value, listType)
        }
    }

    @TypeConverter
    fun estimateListToString(estimateList: List<Estimate>?): String {
        if (estimateList == null) {
            return ""
        } else {
            val gson = Gson()
            return gson.toJson(estimateList)
        }
    }

    @TypeConverter
    fun stringToFareList(value: String?): List<Fare> {
        if (value == null || value == "") {
            return ArrayList()
        } else {
            val listType = object : TypeToken<List<Fare>>() {

            }.type
            return Gson().fromJson(value, listType)
        }
    }

    @TypeConverter
    fun fareListToString(fareList: List<Fare>?): String {
        if (fareList == null) {
            return ""
        } else {
            val gson = Gson()
            return gson.toJson(fareList)
        }
    }

    @TypeConverter
    fun stringToTimestamp(value: String) : Timestamp {
        val listType = object : TypeToken<Timestamp>() {}.type
        return Gson().fromJson(value, listType)
    }

    @TypeConverter
    fun timestampToString(timestamp: Timestamp) : String {
        if(false) {
            return ""
        } else {
            val gson = Gson()
            return gson.toJson(timestamp)
        }
    }
}
